src/libraries/UniswapV3Lib.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.21;
+
+import { IERC20 } from "openzeppelin-contracts/contracts/interfaces/IERC20.sol";
+import { IERC20Metadata } from "openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+
+import { IALMProxy }   from "../interfaces/IALMProxy.sol";
+import { IRateLimits } from "../interfaces/IRateLimits.sol";
+
+import { RateLimitHelpers } from "../RateLimitHelpers.sol";
+
+import { FullMath } from "lib/dss-allocator/src/funnels/uniV3/FullMath.sol";
+import { UniV3Utils } from "lib/dss-allocator/test/funnels/UniV3Utils.sol";
+
+import { ISwapRouter, IUniswapV3PoolLike, INonfungiblePositionManager } from "../interfaces/UniswapV3Interfaces.sol";
+
+import { TickMath } from "lib/dss-allocator/src/funnels/uniV3/TickMath.sol";
+import { UniswapV3OracleLib } from "./UniswapV3OracleLib.sol";
+
+library UniswapV3Lib {
+    uint24 public constant MAX_TICK_DELTA = 887272; // From https://github.com/sky-ecosystem/dss-allocator/blob/dev/src/funnels/uniV3/TickMath.sol#L15
+
+    /**********************************************************************************************/
+    /*** Structs                                                                                ***/
+    /**********************************************************************************************/
+    struct UniswapV3PoolParams {
+        uint24 swapMaxTickDelta;
+        uint32 swapTwapSecondsAgo;
+    }
+
+    struct UniV3Context {
+        IALMProxy   proxy;
+        IRateLimits rateLimits;
+        bytes32     rateLimitId;
+        address     pool;
+        uint256     deadline;
+    }
+
+    struct SwapParams {
+        UniswapV3PoolParams poolParams;
+        address             router;
+        address             tokenIn;
+        uint256             amountIn;
+        uint256             minAmountOut;
+        uint24              tickDelta; // The maximum that the tick can move by after completing the swap; type(uint24).max for no limit
+        uint256             maxSlippage;
+    }
+
+    struct SwapCache {
+        uint160            sqrtPriceLimitX96;
+        uint256            twapExpectedOut;
+    }
+
+    /**********************************************************************************************/
+    /*** External functions                                                                     ***/
+    /**********************************************************************************************/
+
+    // Rate limit decreased by value of token1 
+    function swap(UniV3Context calldata context, SwapParams calldata params) external returns (uint256 amountOut) {
```

Comment:
Maybe we could stick to just one lib input struct, like with most libs? I know it would be pretty big, but still

Reply:
If anything maybe have one struct with stuff loaded from the controller state and the other with what user provided? Then one is config and the other is input/params/sth